### PR TITLE
Surface heap-pressure discovery skips as structured discovery_complete outcome (#2737)

### LINKED ISSUE #2737: Discovery analysis aborted at heap CRITICAL during evolveDir even when run continues
## Summary

During `Creature.evolveDir()` on large supervised `.bin` directories, NEAT-AI's memory monitor frequently enters **CRITICAL** heap state. Discovery passes are **aborted** (`analysis aborted: heap CRITICAL at extension boundary`) while evolution continues — silently reducing the structural search that `evolveDir` is supposed to provide.

## Repro

Any large-record supervised evolveDir run; observed consistently in the MNIST showcase:

```bash
NEAT_EXAMPLES_MAX_HEAP_MB=12288 ./mnist_classification/run.sh --fresh --hidden-seed --timeout=15
# Also with MNIST_HIDDEN_LAYER_SIZES=64 (smaller seed, still hits CRITICAL)
```

## Observed log pattern (NEAT-AI 5.0.29)

```
[MemoryMonitor] Heap: 8724 MB / 9233 MB (94.5%) [CRITICAL]
[Neat] Discovery … analysis aborted: heap CRITICAL at extension boundary
[MemoryMonitor] Critical-level response: cleared all WASM caches, activation cap reduced to 1, compilation cache cleared
```

This repeats across generations. Discovery sometimes completes synapse analysis with "found no improvements" after partial recording timeouts.

## Expected

When Discovery is a core part of the `evolveDir` contract (structural growth + fine-tuning), the library should:

1. **Budget** discovery memory so analysis can complete on realistic workloads, or
2. **Surface** a clear milestone / warning that structural search was skipped due to memory pressure (not just log lines), or
3. **Adapt** discovery batch size / recording scope to available heap instead of aborting.

Example consumers cannot fix this in application code — it is internal to NEAT-AI + NEAT-AI-Discovery coordination.

## Impact on showcase narrative

Examples are meant to demonstrate noise → competent evolution via NEAT-AI's full pipeline. When Discovery aborts under memory pressure, users may see flat topology (794 neurons) and falsely conclude Discovery "doesn't add hidden neurons" when the library actually prevented the analysis pass from finishing.

### PR DESCRIPTION
**Pull request by VibeCoderST**
# Surface heap-pressure discovery skips as a structured training event

## Summary

`DataRecorder.recordFiles()` already aborts the analysis phase when the V8
heap is at MemoryMonitor CRITICAL pressure at the analysis-extension boundary
(PR #2594), and #2642 reduced peak analysis heap. The remaining gap reported
in #2737 is **observability**: the abort emits a single `warn` log line, but
the upstream `discovery_complete` training event still reports `"no_change"`,
so callers can't distinguish a memory-pressure skip from a clean "found
nothing" pass — and they keep concluding Discovery "doesn't add hidden
neurons" when in fact the library prevented the analysis from running.

This PR surfaces the skip as a structured, machine-readable signal end-to-end:

- `DiscoverResult` carries a new optional `heapAbortedAtExtensionBoundary`
  flag. `DataRecorder.recordFiles()` sets it on the empty result it returns
  when the heap guard trips.
- `buildDiscoverResponsePayload` in the worker copies the flag across the
  worker/parent thread boundary so the parent thread sees the same signal as
  the worker thread. `ResponseData.discover` adds the same field.
- `DiscoveryCompleteEvent.outcome` now includes a `"heap_critical_skip"`
  member alongside `"improved" | "no_change" | "timeout"`.
- A small pure helper `chooseDiscoveryCompleteOutcome` in
  `src/NEAT/DiscoveryOutcome.ts` centralises the precedence rule: heap-abort
  wins over `"improved"` and `"no_change"` so a stale
  `improvedCreature` reference can't mask a memory-pressure skip.
- `processCompletedResults` calls the helper instead of inlining the
  outcome ternary.

Callers wiring `onTrainingEvent` can now `switch` on the outcome and surface
the skip as a milestone in their UI / dashboards / showcase narrative,
addressing the second of the three options listed in the issue's "Expected"
section.

Closes #2737.

## Evidence

This is a backend/library change with no UI surface. Behaviour is verified by
deterministic unit tests on the pure outcome helper, the worker-payload
mapper, and the existing heap-guard tests (unchanged and still passing).

```mermaid
sequenceDiagram
    participant DR as DataRecorder.recordFiles
    participant Guard as checkAnalysisHeapAbort
    participant WP as buildDiscoverResponsePayload
    participant PCR as processCompletedResults
    participant Cb as onTrainingEvent

    DR->>Guard: sample heap at extension boundary
    Guard-->>DR: { abort: true } (heap CRITICAL)
    Note over DR: set heapAbortedAtExtensionBoundary = true<br/>on the empty DiscoverResult
    DR-->>WP: DiscoverResult
    WP-->>PCR: ResponseData.discover (flag propagated)
    PCR->>PCR: chooseDiscoveryCompleteOutcome(...)
    Note over PCR: heapAbortedAtExtensionBoundary === true<br/>=> outcome = "heap_critical_skip"
    PCR-->>Cb: { kind: "discovery_complete",<br/>outcome: "heap_critical_skip", ... }
```

Targeted test run:

```
$ deno test --no-check --allow-all \
    test/multithreading/WorkerProcessor.ts \
    test/NEAT/DiscoveryOutcome.ts \
    test/ErrorGuidedStructuralEvolution/AnalysisHeapGuard.ts \
    test/NEAT/TrainingEventEmitter.ts \
    test/ErrorGuidedStructuralEvolution/AnalysisCacheRelease.ts
ok | 33 passed | 0 failed (1s)
```

### Why scope stops here

Of the three options the issue offers — **budget**, **surface**, or **adapt** —
this PR delivers **surface**. Budgeting and adaptive scope reduction require
deeper changes to the recording-phase batch sizing and to NEAT-AI-Discovery's
FFI cache lifecycle, and would not fit the "minimal scope" rule for one
issue. With the structured outcome in place, callers (and a follow-up
adaptive-recording PR) can now measure how often the skip fires and gate
adaptive logic on that signal.

### Pre-existing quality.sh failures

`./quality.sh` currently fails the type-check step on three pre-existing
`setTimeout` typing errors that are unrelated to this PR:

- `src/architecture/ErrorGuidedStructuralEvolution/DataRecorderAnalysis.ts:576`
- `src/architecture/ErrorGuidedStructuralEvolution/DataRecorderRecording.ts:371`
- `src/intelligentDesign/ImproveSquash.ts:43`

Verified by stashing this PR's changes and re-running `./quality.sh
--check-only` — the same three errors fire on `origin/Develop`. Per the
project's "Change Scope" rule these are left for a separate fix; the new
code in this PR type-checks cleanly in isolation
(`deno check src/NEAT/DiscoveryOutcome.ts` passes).

## Test Plan

- Added `test/NEAT/DiscoveryOutcome.ts`:
  - `chooseDiscoveryCompleteOutcome: no inputs → no_change`.
  - `chooseDiscoveryCompleteOutcome: improvedCreature present → improved`.
  - `chooseDiscoveryCompleteOutcome: heap abort wins over no improvement →
    heap_critical_skip`.
  - `chooseDiscoveryCompleteOutcome: heap abort wins over improvedCreature
    → heap_critical_skip` (defends precedence even if both signals appear).
  - `chooseDiscoveryCompleteOutcome: heapAbortedAtExtensionBoundary=false
    preserves the regular precedence`.
- Extended `test/multithreading/WorkerProcessor.ts`:
  - `buildDiscoverResponsePayload: propagates heapAbortedAtExtensionBoundary
    =true` — confirms the flag survives the worker→parent wire copy.
  - `buildDiscoverResponsePayload: leaves heapAbortedAtExtensionBoundary
    undefined for happy-path results`.
  - `buildDiscoverResponsePayload: propagates heapAbortedAtExtensionBoundary
    =false`.
- Existing `test/ErrorGuidedStructuralEvolution/AnalysisHeapGuard.ts` and
  `test/NEAT/TrainingEventEmitter.ts` continue to pass — the guard's abort
  signal and the discovery_complete event delivery path are unchanged apart
  from the new outcome member.

---

🤖 Processed by: VibeCoderST

### CODE REVIEW

**Comment by VibeCoderST** [comment]
## CI Fix — `setTimeout` type errors resolved

The `quality` (and `coverage`) checks were failing with three `TS2322` errors:

```
Type 'Timeout' is not assignable to type 'number'.
```

### Root cause

Deno's `setTimeout` returns an opaque `Timeout` object, not a plain `number`. Three timeout handle variables were declared as `number | undefined`, which was incompatible in Deno's type checker.

### Fix

Changed all three declarations from `number | undefined` to `ReturnType<typeof setTimeout> | undefined`:

| File | Line |
|------|------|
| `src/architecture/ErrorGuidedStructuralEvolution/DataRecorderAnalysis.ts` | 574 |
| `src/architecture/ErrorGuidedStructuralEvolution/DataRecorderRecording.ts` | 368 |
| `src/intelligentDesign/ImproveSquash.ts` | 41 |

### Verification

`./quality.sh --check-only` exits 0 locally after the fix. The type errors no longer appear.

**Comment by VibeCoderST** [comment]
## CI Failure Investigation: Analyze (javascript-typescript)

### What was inspected

- PR #2738 check runs via `gh pr checks`
- Failed job details: run `26329400600`, job `77512730466` ("Analyze (javascript-typescript)")
- Job step breakdown: the failure occurs at **"Checkout repository"** (step 2); all subsequent CodeQL steps were skipped
- Two CodeQL runs on this PR's merge ref (`refs/pull/2738/head`):
  - Run `26329397344` (commit `383d193`) — **passed** ✓
  - Run `26329400600` (commit `f05ce37`) — **failed** ✗
- The only difference between those commits is `chore: auto-increment patch version` — no workflow or CI configuration changed
- All other checks on commit `f05ce37` (Quality reviews, ShellCheck, Semgrep, Dependency review, Markdown Lint, Spelling) **passed**

### Root cause

The failure is in **GitHub's managed dynamic CodeQL workflow** (`dynamic/github-code-scanning/codeql`, workflow ID 132372010), not in any custom workflow file in this repository. The checkout step failed with:

```
could not read Username for 'https://github.com': terminal prompts disabled
The process '/usr/bin/git' failed with exit code 128
```

This is a git credential error during the CodeQL runner's checkout of the PR merge commit. The same code (commit `383d193`) passed CodeQL successfully 11 seconds earlier, confirming this is not a code-quality issue introduced by the PR.

### Why no code change was applied

There is no in-repository code or configuration to change. The dynamic CodeQL workflow is managed entirely by GitHub and is not configurable via workflow YAML files in this repo. The error is a runner-side authentication failure unrelated to the PR's TypeScript changes.

An API re-run attempt returned HTTP 403 (`This workflow run cannot be retried`) — dynamic workflows can only be re-triggered by pushing a new commit or by a maintainer clicking **Re-run failed jobs** in the GitHub Actions UI.

### Recommended next steps

1. **Re-run the failed check**: A repository maintainer can navigate to the [failed run](https://github.com/stSoftwareAU/NEAT-AI/actions/runs/26329400600) and click **Re-run failed jobs**. The same code passed CodeQL on the earlier commit, so a clean re-run is expected to succeed.
2. **Alternatively, push a trivial commit** (e.g. a whitespace change) to the PR branch to trigger a fresh CodeQL run.
3. If the re-run also fails, investigate whether the repository's CodeQL permissions (`security-events: write`, `actions: read`) are correctly configured in the dynamic workflow settings.

**Comment by VibeCoderST** [comment]
I investigated the CI check failure (**Analyze (javascript-typescript)**) but could not determine a fix — please review.

**Classifier reason:** no recognised pattern in check 'Analyze (javascript-typescript)' or its annotations
**Signals:**
- `check:analyze (javascript-typescript)`

**Comment by VibeCoderST** [comment]
I investigated the CI check failure (**Analyze (javascript-typescript)**) but could not determine a fix — please review.

**Classifier reason:** no recognised pattern in check 'Analyze (javascript-typescript)' or its annotations
**Signals:**
- `check:analyze (javascript-typescript)`

**Review by nleck — APPROVED** [review]
